webhook/pr_review.go

```diff
@@ -0,0 +1,374 @@
+			if err != nil {
+				log.Printf("Error submitting review comment for file %s: %v\n", file.Filename, err)
+			} else {
+				log.Printf("Review comment submitted for file %s\n\n", file.Filename)
+			}
+			builder.Reset()
+		}
+	}
+}
+
+func getChatGPTComment(patch string) (string, error) {
+	client, err := openai.New(
+		openai.WithToken(viper.GetString("openai.api_key")),
+		openai.WithModel(viper.GetString("openai.model")),
+		openai.WithOrgID(viper.GetString("openai.org_id")),
+		openai.WithProxyURL(viper.GetString("openai.proxy")),
+		openai.WithSocksURL(viper.GetString("openai.socks")),
+		openai.WithBaseURL(viper.GetString("openai.base_url")),
+		openai.WithTimeout(viper.GetDuration("openai.timeout")),
+		openai.WithMaxTokens(viper.GetInt("openai.max_tokens")),
+		openai.WithTemperature(float32(viper.GetFloat64("openai.temperature"))),
+	)
+	if err != nil {
+		log.Println("Error creating OpenAI client:", err)
+		return "", err
+	}
+
+	out, err := util.GetTemplateByString(
+		prompt.CodeReviewTemplate,
+		util.Data{
+			"file_diffs": patch,
+		},
+	)
+	if err != nil {
+		log.Println("Error creating prompt:", err)
+		return "", err
+	}
+
+	ctx := context.Background()
+	resp, err := client.Completion(ctx, out)
+	if err != nil {
+		log.Println("Error completing prompt:", err)
+		return "", err
+	}
+	summarizeMessage := resp.Content
+
+	log.Println("PromptTokens: " + strconv.Itoa(resp.Usage.PromptTokens) +
+		", CompletionTokens: " + strconv.Itoa(resp.Usage.CompletionTokens) +
+		", TotalTokens: " + strconv.Itoa(resp.Usage.TotalTokens))
+
+	// Output core review summary
+	log.Println("================Review Summary====================")
+	log.Println("\n" + strings.TrimSpace(summarizeMessage) + "\n\n")
+	log.Println("==================================================")
+
+	return summarizeMessage, nil
+}
+
+func getLastLinePosition(patch string) int {
+	lines := strings.Split(patch, "\n")
+	position := 0
+	for _, line := range lines {
+		if strings.HasPrefix(line, "@@") {
+			continue
+		}
+		position++
+	}
+	return position
+}
+
+func submitReviewComment(owner, repo string, prNumber int, commitID, filePath, comment string, position int) error {
+	client := &http.Client{}
+
+	commentRequest := CommentRequest{
+		Body:     comment,
+		Path:     filePath,
+		Position: position,
+		CommitID: commitID,
+	}
+
+	jsonData, err := json.Marshal(commentRequest)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest("POST", fmt.Sprintf(submitCommentsURL, apiBaseurl, owner, repo, prNumber), bytes.NewBuffer(jsonData))
+	if err != nil {
+		log.Println("Error creating request:", err)
+		return err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", "token "+token)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Println("Error submitting review comment:", err)
+		return err
+	}
+	defer resp.Body.Close()
+
```

Comment:
Overall, the code patch looks fine, but there are some areas for improvement:

1. Error handling: The error handling in the code patch is minimal, and it may cause issues if any errors occur during runtime. It is recommended to add more error handling to catch and handle potential errors.

2. Security vulnerabilities: The code patch does not appear to have any security vulnerabilities, but it is always recommended to review the code for potential security issues.

3. Code readability: The code patch could be improved by adding comments and making the code more readable. This will make it easier for other developers to understand the code and make changes when necessary.

4. Code optimization: The code patch has some areas that could be optimized, such as reducing the number of API calls or improving the performance of certain functions. These optimizations can help improve the overall performance of the code.

5. Testing: The code patch does not include any unit tests or integration tests, which can make it difficult to identify and fix bugs. It is recommended to add testing to ensure that the code works as expected.